parser/impl/parser.cpp

```diff
@@ -469,17 +469,201 @@ sPtr<ast::IfStatement> Parser::parseIfStatement() {
+}
+
+bool Parser::isPrimitive(TokenType token) {
+    return token == TokenType::Identifier || token == TokenType::Int ||
+           token == TokenType::Real || token == TokenType::True ||
+           token == TokenType::False;
+}
+
+sPtr<ast::Expression> Parser::parseUnaryExpression() {
+    Token tok = m_lexer.Peek();
+    ast::Expression expr;
+    if (opPrec(tok.type) >= 0) {
+        if (tok.type == TokenType::Not || tok.type == TokenType::Sub ||
+            tok.type == TokenType::Add) {
+
+            tok = m_lexer.Next();
+            expr.operation = tok;
+            expr.operand1 = parseUnaryExpression();
+            return std::make_shared<ast::Expression>(expr);
+
+        } else if (tok.type == TokenType::OpenParen) {
+            tok = m_lexer.Next();
+            sPtr<ast::Expression> expr = parseBinaryExpression(0);
+            if (m_lexer.Peek().type == TokenType::CloseParen) {
+                tok = m_lexer.Next();
+                return expr;
+            }
+            // error
+            m_errors.push_back(Error{
+                .pos = tok.pos,
+                .message = "Expected to find ')'",
+            });
+            return nullptr;
+        } else {
+            // throw error
+            m_errors.push_back(Error{
+                .pos = tok.pos,
+                .message = "Expected to find unary operator",
+            });
+            return nullptr;
+        }
+    } else if (isPrimitive(tok.type)) {
+        tok = m_lexer.Next();
+        // check if More Then Identifier
+        if (tok.type == TokenType::Identifier) {
+
+            Token t = m_lexer.Peek();
+            if (t.type == TokenType::OpenParen) {
+                // parse routineCall:
+                ast::RoutineCall rc;
+                rc.routine = tok; // save the function name
+                expr.operation = t;
+                do {
+                    t = m_lexer.Next(); // first call-read '(', others - ','
+                    // read expression
+                    sPtr<ast::Expression> e = parseBinaryExpression(0);
+                    // append to the vector
+                    rc.args.push_back(e);
+
+                } while (m_lexer.Peek().type == TokenType::Comma);
+
+                if (m_lexer.Peek().type != TokenType::CloseParen) {
+                    m_errors.push_back(Error{
+                        .pos = m_lexer.Peek().pos,
+                        .message = "Expected to find ')'",
+                    });
+                    return nullptr;
+                }
+                t = m_lexer.Next(); // read ')'
+                expr.operand1 = std::make_shared<ast::RoutineCall>(rc);
+                return std::make_shared<ast::Expression>(expr);
+
+            } else if (t.type == TokenType::Dot ||
+                       t.type == TokenType::OpenBrack) {
+                // we can have a.b.c[7+9].d[0].a
+                // Identifier { . Identifier | [ Expression ] }
+                ast::ModifiablePrimary mp;
```

Comment:
What about a separate `parseModifiablePrimary` function?

Reply:
I haven't done it because in this place I have consumed the first Identifier to reach the next token. Thus we'll need additional parameters to `parseModifiablePrimary`, which will be out of structure.

Reply:
We have consumed the first identifier to find if the sequence of tokens represent `ModifiablePrimary`, thus we need to pas such values as an arguments, which by the nature is not a usual `parse...` function.